File: java/Lynda.com Up and Running with Java - notes and code/05_beyond_the_basics/sorting_arrays/InsertionSort.java
// Sorts in descending order
public class InsertionSort {
  
  public static void main(String[] args) {
    
    int[] nums = {54, 86, 2, 12, 34, 65, 75, 6, 44};

    insertionSort(nums);

    for (int i = 0; i < nums.length; i++) {
      System.out.print(nums[i]+"\t");
    } 
  }

  public static void insertionSort(int[] arr) {
    int j, key, i;
    for (j = 1; j < arr.length; j++) {
      key = arr[j];
      for (i = j - 1; (i >= 0) && (arr[i] < key); i--) {
        arr[i + 1] = arr[i];    
      }

      arr[i + 1] = key;
    }

  }

}

File: algorithms/javavscript/sort/insertion_sort.js
// Insertion Sort
// Starting with a unsorted list, also create an empty "sorted" list.
// Start at the first number of the unsorted list, it goes into the sorted list straight away because there are no other elements there
// Each element in the unsorted list gets compared to each element in the sorted list from left to right
// If the unsorted element is lower than the sorted element the sorted element moves one place to the right

// Should implement one element at a time in its correct place

var arr = [1000, 4, 2, 13, 3, 2, 1];

function insertionSort(nums) {
  for (let i = 1; i < nums.length; i++) {
    for (let j = 0; j < i; j++) {
      if (nums[i] < nums[j]) {
        let spliced = nums.splice(i, 1)
        nums.splice(j, 0, spliced[0])
        break
      }
    }
  }

  return nums
}

console.log(insertionSort(arr))

// O(n²) 
// Better than bubble sort because it doesn't go through the entire array 
// twice the inner loop only goes through part of the array

